src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/ReferenceCleanupService.cs

```diff
@@ -87,47 +93,45 @@ private static async Task<List<ProjectSystemReferenceInfo>> GetAllReferencesInCo
             return references;
         }
 
-        public async Task<bool> TryUpdateReferenceAsync(string projectPath, ProjectSystemReferenceUpdate referenceUpdate, CancellationToken cancellationToken)
+        public Task<bool> TryUpdateReferenceAsync(string projectPath,
+            ProjectSystemReferenceUpdate referenceUpdate, CancellationToken cancellationToken)
         {
-            bool wasUpdated = false;
+            throw new NotImplementedException();
+        }
 
+        /// <summary>
+        /// Gets an operation that can update the project’s references by removing or marking references as
+        /// TreatAsUsed in the project file.
+        /// </summary>
+        public async Task<IProjectSystemUpdateReferenceOperation> GetUpdateReferenceOperationAsync(string projectPath, ProjectSystemReferenceUpdate referenceUpdate,
+            CancellationToken cancellationToken)
+        {
             cancellationToken.ThrowIfCancellationRequested();
 
             var activeConfiguredProject = await GetActiveConfiguredProjectByPathAsync(projectPath, cancellationToken);
 
             var referenceHandler = s_mapReferenceTypeToHandler[referenceUpdate.ReferenceInfo.ReferenceType];
 
+            IProjectSystemUpdateReferenceOperation command = s_nullCommand;
             if (referenceHandler != null)
             {
-                wasUpdated = await ApplyActionAsync(referenceUpdate, referenceHandler, activeConfiguredProject, cancellationToken);
+                command = CreateCommand(referenceUpdate, referenceHandler, activeConfiguredProject, cancellationToken);
```

Comment:
>command [](http://example.com/codeflow?start=16&length=7)

minor style observation you can ignore: this local variable isn't needed if you use a return statement here.

Reply:
Actually I wanted to have only one exit point instead of having many returns.

Reply:
That's fair, but it's not a commonly followed pattern in this repo (or in .NET in general).